index.d.ts

```diff
@@ -0,0 +1,490 @@
+  }
+}
+interface Monad<A> extends Applicative<A>, Chain<A> {}
+interface Alt<A> extends Functor<A> {
+  'fantasy-land/alt': (other: Alt<A>) => Alt<A>
+}
+interface PlusTypeRep<A> {
+  'fantasy-land/zero': () => Plus<A>
+}
+interface Plus<A> extends Alt<A> {
+  constructor: PlusTypeRep<A>
+}
+interface Alternative<A> extends Applicative<A>, Plus<A> {
+  constructor: ApplicativeTypeRep<A> & PlusTypeRep<A>
+}
+interface Foldable<A> {
+  'fantasy-land/reduce': <B>(f: (B, A) => B, x: B) => B
+}
+interface Traversable<A> extends Functor<A>, Foldable<A> {
+  'fantasy-land/traverse': <B>(typeRep: TypeRep, f: (A) => Applicative<B>) => Applicative<Traversable<B>>
+}
+interface Extend<A> extends Functor<A> {
+  'fantasy-land/extend': <B>(f: ($0: Extend<A>) => B) => Extend<B>
+}
+interface Comonad<A> extends Extend<A> {
+  'fantasy-land/extract': () => A
+}
+interface Contravariant<A> {
+  'fantasy-land/contramap': <B>(f: (B) => A) => Contravariant<B>
+}
+
+
+export const Maybe: TypeRep;
+export const Nothing: Maybe<any>;
+export function Just<A>(value: A): Maybe<A>;
+
+export const Either: TypeRep;
+export function Left<A, B>(value: A): Either<A, B>;
+export function Right<A, B>(value: B): Either<A, B>;
+
+//  Classify
+
+export function type(x: any): {
+  namespace: Maybe<string>
+  name: string
+  version: NonNegativeInteger
+};
+
+export function is(typeRep: TypeRep, x: any): boolean;
+export function is(typeRep: TypeRep): (x: any) => boolean;
+
+//  Showable
+
+export function toString(x: any): string;
+
+//  Fantasy Land
+
+export function equals<A>(x: null, y: null): boolean;
+export function equals<A>(x: null): (y: null) => boolean;
+export function equals<A>(x: undefined, y: undefined): boolean;
+export function equals<A>(x: undefined): (y: undefined) => boolean;
+export function equals<A>(x: boolean, y: boolean): boolean;
+export function equals<A>(x: boolean): (y: boolean) => boolean;
+export function equals<A>(x: number, y: number): boolean;
+export function equals<A>(x: number): (y: number) => boolean;
+export function equals<A>(x: Date, y: Date): boolean;
+export function equals<A>(x: Date): (y: Date) => boolean;
+export function equals<A>(x: RegExp, y: RegExp): boolean;
+export function equals<A>(x: RegExp): (y: RegExp) => boolean;
+export function equals<A>(x: string, y: string): boolean;
+export function equals<A>(x: string): (y: string) => boolean;
+export function equals<A>(x: Array<A>, y: Array<A>): boolean;
+export function equals<A>(x: Array<A>): (y: Array<A>) => boolean;
+export function equals<A>(x: IArguments, y: IArguments): boolean;
+export function equals<A>(x: IArguments): (y: IArguments) => boolean;
+export function equals<A>(x: Error, y: Error): boolean;
+export function equals<A>(x: Error): (y: Error) => boolean;
+//  XXX: This is too general. Can we match "plain" objects only?
+//  export function equals<A>(x: Object, y: Object): boolean;
+//  export function equals<A>(x: Object): (y: Object) => boolean;
```

Comment:
There's `object` and `{}`. I haven't even bothered to look into their nuances. `Object` appears not used as it'd match almost anything.

On DRY'ing... `equals<A>(x: any, y: any): boolean`? 😅
Similar to the `lt` / `gt`; wanted to say there's `equals<A extends null | undefined | boolean | ...>(x: A, y: A): boolean` (or even drop the constraint), but... generics don't just fail with literals here, it just does nothing useful. At that point it'll just be like, `equals(123, "foo")` totally works for `A`: `any` (or for `number | string`).
So technically without run-time type checks `equals(123, "foo")` may still run and produce false I guess?
I'd wanna go for a type-level type check, but those aren't possible yet. Failing that, it's your solution or any. If you'd consider not doing run-time errors for different types, with your approach you might even add a `equals<A>(x: any, y: any): false` fallback at the bottom.